### PR TITLE
Update the Dependency Management System (DMS)

### LINKED ISSUE #330: Create an Eradiate package [2/3]
This feature request is the second step in the creation of a PyPI Eradiate package. This issue is created to discuss the requirements and possible implementations of a patched [dependencies management system](https://eradiate.readthedocs.io/en/stable/rst/maintainer_guide.html#managing-dependencies) (DMS) suitable for the wheel distribution. A few design ideas have been developed, and are left here for the record. The design number 3 is the one implemented in this PR. This PR is the second in a series of 3 PRs related to the packaging. The first one #328 focused on the implementation of a wheel for Eradiate and its upload to PyPI, and the last one will focus on the release process, testing of the different envs, and documentation.

## Summary

Since #328 , an Eradiate package is [available through Pypi](https://pypi.org/project/eradiate/). It can be installed alongside the Eradiate-specific Mitsuba rendering kernel compiled for that purpose using the following pip command:

```
pip install "eradiate[production]"
```

The `[production]` suffix bothers us very much, because using the default, simpler, and more intuitive command `pip install eradiate` would not install Mitsuba, and thus provide an incomplete installation to end users. 

## Reason and benefit

We could stick to this state of affairs and gently suggest anyone who do not install the package correctly to RTFM, but I am a depressive maniac, not a mean developer. As a result, we need to find a solution to this *cosmetic* problem.

On top of this inconvenience, the current DMS forces the wheel to define the `dev`, `tests`, `docs`, and `optional` dependency layers as optional dependency sets. These sets are defined for development or CI/CD purposes only, for use cases where having access to an Eradiate source repository is mandatory. As such, including them in the Eradiate wheel definition, only used for production by end users working outside of the sources, is an anti-pattern and might only produce confusion.

Furthermore, no convenient way to switch deps layer has been provided to developers so far, to the best of my knowledge. This could be a nice to have feature.

## Details

The following subsections are descriptions of the different setups we thought about and their issues.

### Design 0

We stick to the current setup as of #328 . Eradiate installation is documented for prod with highlights on the importance of using the `eradiate[production]` deps set. It is simple. It is ready, we can move on to other topics. It is ugly. Some users will fail to see the doc. It's a bit weird to have a broken install when simply doing `pip install eradiate`.

### Design 1

We add eradiate_mitsuba to the main layer. Hence it is installed anyway in any setup. If an Eradiate source repository is detected by the eradiate *code*, then it has the precedence over the Eradiate package **and** the Eradiate Mitsuba kernel package.

The problems with this setup are the following:
- The mitsuba command precedence can not be defined by the wheel and will depend on the user environment. As a result, we can not guarantee which binary would be called when a developer performs a CLI `mitsuba` operation: the packaged one or the one built in source?
- It won't be possible to have "light-weight" layers not requiring Mitsuba for some specific CI/CD steps for instance.
- It breaks the deps layering concept in the first place. There is no good reason to have layers if we then have special cases for some packages
- The wheel uses the optional dependencies as defined in the `pyproject.toml`. We would thus end up with inconsistent layer definitions between the unused dev dep sets defined in the wheel, and the actual dev deps layers used when working on the Eradiate source tree.
- I am not sure if it's easily maintainable or not

### Design 2

We deploy two different packages to Pypi:
- `eradiate-dev`
- `eradiate`

`eradiate-dev` is the renamed package implementation as of #328 . `eradiate` is just a shallow package, requiring `eradiate-dev[production]`.

It's rather simple to implement, and (only) fixes the *cosmetic* issue.

Problems:
- It will create some confusion.
- Uninstalling `eradiate` through `pip uninstall eradiate` would not uninstall `eradiate-dev`, and thus the Eradiate actual implementation would still be available after uninstalling
- It does not solve the problems of dev deps being uploaded to Pypi as optional deps of the prod package.

### Design 3

I tend to believe now that a lot of the problems we see RN are due to the fact we define the layers in the pyproject.toml file as optional dependencies. The pyproject.toml file is here to define the deps of our production wheel. I don't see real good reasons to have the dev layers defined there in the end.

Since we already have a set of script that generate `requirements*.txt` files in the `requirements` directory, we could rewrite our pyproject.toml to include those latter instead. This is fairly easy to do using the [`dynamic` field of the pyproject.toml](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata), and I have succeeded in doing so while testing a bit. The `requirements/*.py` scripts read the pyproject.toml optional dependency sets in the first place, so we would indeed need to reformulate them to source their informations on the layers from somewhere else.

The idea here is to redefine the layers of the DMS in another file than `pyproject.toml`. The scripts in the requirements folder would then read the layers definitions and create requirement-*.txt files. Finally, the `pyproject.toml` only imports the requirement lists necessary for its default and optional dependency sets. Extending the `requirements/layered.yml` would be a good candidate for defining the content of the source layers.

## Additional improvements

~As of #328 , conda deps are assumed to be transparently translatable from Pypi package names. This is not really the case, and we could instead generate conda environments that add a pip deps if the exact same conda package can not be found. Installing pip deps using conda env definitions is supported by Conda. This is the case in particular for the eradiate_mitsuba package that does not have a conda counterpart. As a result, #328 provides a simple fix for this specific package. The idea here would be to generalize this process.~ Will not fix: we prefer to handle these cases manually to avoid shipping incorrect deps without any warning.

~Dependency layers production and packaging requires a bit more testing in the CI.~ Out of the scope of the current PR. Should be done in subsequent PR (3/3) updating the CI, env testing, and release process.

Nice to have: provide a simple way to switch layer in dev mode

### PR DESCRIPTION
**Pull request by wint3ria**
Closes #330 

# Description

Modify the DMS according to #330 , design 3:

- Decouple the DMS configuration from setuptools configuration file (`pyproject.toml`)
- Modify the dependency layer graph (see `graph.png` file). Rename `main` to `dependencies` for naming consistency with `pyproject.toml` 

graph

- Use `networkx` to handle the layer graph in the conda env generation script
- Import the pip requirement files for the `dependencies` and `recommended` layers in `pyproject.toml`. Only these two sets are thus exposed by the wheel as the `dependencies` list and a `optional-dependencies` list respectively.
- Move the deps definitions for all layers to the `requirements/layered.yml` file. This file now defines the graph and the required packages for each node/layer.
- Move the relevant pip and conda configurations to dedicated folders requirements/pip and requirements/conda
- Allow to output conda deps graph in dot format for debugging

## Layers

**main**: minimal requirements for eradiate to run in dev mode
**depencencies**: dependencies list used by default by setuptools in production packages. Includes the eradiate-mitsuba pkg
**recommended**: convenient optional dependencies included in the production package. Installable through `pip install eradiate[recommended]`
**tests**: required for testing eradiate in dev mode
**docs**: required to compile the docs in dev mode
**dev**: dev setup, includes docs, tests and recommended
**optional**: dev setup including the eradiate-mitsuba pkg for convenience

# Note

This PR removes the previously defined pip and conda lock files, and does not provide any new one. Merging this PR will require to recreate the lock files.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I updated the change log if relevant
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

### CODE REVIEW

**Review by leroyvn — COMMENTED** [review]
Hi @wint3ria, this looks good. The only point I'd like to raise is the deletion of lock files. They are checked in to save the pain of solving the environment during setup (very useful on slow CI/CD machines, *e.g.* on ReadTheDocs). What was your motivation when you decided to remove them?

**Comment by leroyvn** [comment]
Regarding your problem with the macOS wheel for Dr.Jit: the wheel you're requesting [seems to exist](https://pypi.org/project/drjit/0.4.1/#files), so it could be that the Poetry solver, used by conda-lock for pip dependencies, just can't find it (pretty embarrassing since Dr.Jit simply doesn't have dependencies). Could worth a report to the conda-lock folks.

**Comment by wint3ria** [comment]
> Hi @wint3ria, this looks good. The only point I'd like to raise is the deletion of lock files. They are checked in to save the pain of solving the environment during setup (very useful on slow CI/CD machines, _e.g._ on ReadTheDocs). What was your motivation when you decided to remove them?

I removed them to avoid conflicts on rebase. The purpose of this PR is to *reshape* the way we create these files. I do not mean that lock files are not versioned anymore. The scripts I updated generate lock files into the `requirements/pip` and `requirements/conda` folders instead of putting everything directly in `requirements`. The naming convention is also slightly different. 

Considering all of this, I thought natural to consider previous lock files as outdated, and that we would recreate the files when we rebase/merge this PR. I did not want to keep outdated lock files in the PR, and likely the lock files that I could generate with the new scripts could be outdated after merging. I can indeed provide new lock files as part of this PR if you prefer.

File: Makefile [hunk 1]
```diff
@@ -54,36 +54,53 @@ pip-update: pip-lock pip-init
...
-	conda update --file requirements/environment-$(PLATFORM).lock
+
+no-deps-install:
 	python3 requirements/copy_envvars.py
 	pip install --editable . --no-deps
 
+# Initialise development environment
+conda-init: conda-prepare
+	conda update --file requirements/conda/environment-dev-$(PLATFORM).lock
+	$(MAKE) no-deps-install
```
[the remark this answers is not in the document]
Comment (wint3ria) [reply Makefile:87]: Yes, I agree. I thought this PR in this sense, extending the behavior you described to allow more lock files to be generated for other types of installation (docs and tests in CI/CD for instance)

**Comment by wint3ria** [comment]
> > Hi @wint3ria, this looks good. The only point I'd like to raise is the deletion of lock files. They are checked in to save the pain of solving the environment during setup (very useful on slow CI/CD machines, _e.g._ on ReadTheDocs). What was your motivation when you decided to remove them?
> 
> I removed them to avoid conflicts on rebase. The purpose of this PR is to _reshape_ the way we create these files. I do not mean that lock files are not versioned anymore. The scripts I updated generate lock files into the `requirements/pip` and `requirements/conda` folders instead of putting everything directly in `requirements`. The naming convention is also slightly different.
> 
> Considering all of this, I thought natural to consider previous lock files as outdated, and that we would recreate the files when we rebase/merge this PR. I did not want to keep outdated lock files in the PR, and likely the lock files that I could generate with the new scripts could be outdated after merging. I can indeed provide new lock files as part of this PR if you prefer.

Especially now since we have a problem on osx, we may want for a solution before generating the lock files?

**Comment by wint3ria** [comment]
> Regarding your problem with the macOS wheel for Dr.Jit: the wheel you're requesting [seems to exist](https://pypi.org/project/drjit/0.4.1/#files), so it could be that the Poetry solver, used by conda-lock for pip dependencies, just can't find it (pretty embarrassing since Dr.Jit simply doesn't have dependencies and therefore can be installed anywhere). Could be worth a report to conda-lock folks.
> 
> Edit: I posted an issue on the conda-lock issue tracker ([conda/conda-lock#436](https://github.com/conda/conda-lock/issues/436))

After discussion, it was decided to add an option to the conda env generation script to disable the pip deps (and drjit in particular). This is a temporary workaround, until poetry solver is fixed or we have access to a conda package for mitsuba.

**Comment by wint3ria** [comment]
As a result of the poetry solver issue, use who may want to use the conda environments `optional` or `dependencies` (from the locks or the env yaml files) will need to perform the installation of the `eradiate-mitsuba` pkg on their own using pip. This is a very minor issue, since these conda environments are only provided for rare corner cases. The `dependencies` layer is used for the wheel main deps list, its purpose isn't to be installed this way with conda. The `optional` layer is only a convenience layer, and most likely won't be used. The new option `-p` for `pip-deps` allow the script to generate conda envs with the pip deps when the solver will be fixed. Complete `dependencies` and `optional` environments can be setup using the pip locks anyway.

**Pull request merged**